js/hw_2_js_functions_state/src/game_of_snake/game_of_snake.js

```diff
@@ -1,3 +1,440 @@
+    };
+  }
+
+  return state;
+});
+
+engine.addSignalReducer('pauseGame', (state) => {
+  return {
+    ...state,
+    gameOn: !state.gameOn,
+  };
+});
+
+engine.addSignalReducer('moveSnake', (state) => {
+  const prevSnakeHead = { ...state.snakeHead };
+  let x = prevSnakeHead.x;
+  let y = prevSnakeHead.y;
+
+  switch (state.direction) {
+    case 'moveDown':
+      y++;
+      break;
+    case 'moveUp':
+      y--;
+      break;
+    case 'moveRight':
+      x++;
+      break;
+    case 'moveLeft':
+      x--;
+      break;
+
+    default:
+      break;
+  }
+
+  const nextSnakeHead = {
+    x,
+    y,
+  };
+
+  const minBorderCoords = 0;
+  const maxBorderCoords = 26;
+  const isSnakeHitBorders =
+    x === maxBorderCoords ||
+    y === maxBorderCoords ||
+    x === minBorderCoords ||
+    y === minBorderCoords;
+  const isSnakeHitBody = state.snakeCoords.includes(getIndex(x, y));
+
+  if (isSnakeHitBody || isSnakeHitBorders) {
+    return {
+      ...state,
+      gameLost: true,
+      snakeCoords: [],
+      dishIndexes: [],
+    };
+  }
+
+  const catchDishPoint = 1;
+  const snakeLength = state.snakeCoords.length;
+  const snakeHeadCoordinates = [...state.snakeCoords][snakeLength - 1];
+  const prevStateDishIndexes = [...state.dishIndexes];
+  const indexDishToRemove = prevStateDishIndexes.indexOf(snakeHeadCoordinates);
+  const newData = state.dishIndexes.includes(snakeHeadCoordinates)
+    ? [...state.snakeCoords]
+    : [...state.snakeCoords].slice(1, snakeLength);
+
+  if (indexDishToRemove > -1) {
+    prevStateDishIndexes.splice(indexDishToRemove, 1);
+  }
+
+  return {
+    ...state,
+    snakeHead: nextSnakeHead,
+    currentScore: state.dishIndexes.includes(snakeHeadCoordinates)
+      ? state.currentScore + catchDishPoint
+      : state.currentScore,
+    dishIndexes: state.dishIndexes.includes(snakeHeadCoordinates)
+      ? prevStateDishIndexes
+      : state.dishIndexes,
+    snakeCoords: newData.concat(getIndex(nextSnakeHead.x, nextSnakeHead.y)),
+  };
+});
+
+engine.addSignalReducer('giveNewDish', (state) => {
+  if (getIsGameEnded(state)) {
+    return state;
+  }
+
+  const prevDishIndexes = [...state.dishIndexes];
+  const nextDishIndex = getRandomDishPosition(
+    prevDishIndexes,
+    state.snakeCoords
+  );
+
+  return {
+    ...state,
+    dishIndexes: prevDishIndexes.concat(nextDishIndex),
+    round: state.round + 1,
+  };
+});
+
+engine.start();
+
+function renderGameAlert({ state }, emit) {
+  let alert = document.querySelector('.alert');
+
+  if (!alert) {
+    alert = document.createElement('div');
+    alert.classList.add('alert');
+    alert.innerHTML = 'Game has ended. Press "Enter" to start again';
+    document.body.appendChild(alert);
+  }
+
+  const onAlertEnter = (e) => {
+    if (e.key === 'Enter') {
+      emit('restartGame');
+    }
+  };
+
+  if (state.gameLost) {
+    document.addEventListener('keyup', onAlertEnter);
+    alert.classList.add('alert-loss');
+  } else {
```

Comment:
Technically, we could `return () => { document.removeEventListener... }` right here.